aiohttp/web_ws.py

```diff
@@ -185,12 +186,60 @@ def _ping_task_done(self, task: "asyncio.Task[None]") -> None:
         self._ping_task = None
 
     def _pong_not_received(self) -> None:
+        """Callback for when no PONG was received after self._pong_heartbeat seconds"""
         if self._req is not None and self._req.transport is not None:
-            self._handle_ping_pong_exception(
-                asyncio.TimeoutError(
-                    f"No PONG received after {self._pong_heartbeat} seconds"
+            loop = self._loop
+            if (
+                loop is not None and not self._waiting
+            ):  # If self._waiting is set we already are in the receive loop and would have read the PONG if one was there
+                pong_not_received_task = loop.create_task(
+                    self._pong_not_received_coro()
                 )
+                if not pong_not_received_task.done():
+                    self._pong_not_received_task = pong_not_received_task
+                    pong_not_received_task.add_done_callback(
+                        self._pong_not_received_done
+                    )
+                else:
+                    self._pong_not_received_done(pong_not_received_task)
+            else:
+                self._handle_ping_pong_exception(
+                    asyncio.TimeoutError(
+                        f"No PONG received after {self._pong_heartbeat} seconds"
+                    )
+                )
+
+    async def _pong_not_received_coro(self) -> None:
+        """Coroutine to check for pending PONG when no PONG was received after self._pong_heartbeat seconds"""
+        reader = self._reader
+        assert reader is not None
+        try:
+            async with async_timeout.timeout(self._pong_heartbeat / 10.0):
+                msg = await reader.read()
```

Comment:
Is it possible for a user to catch the exception or anything and avoid the connection close? If it is, then this would break their code, as we've now read a message that they should be processing.

Reply:
I think this is the same point as https://github.com/aio-libs/aiohttp/pull/10544#pullrequestreview-2686358623